File: week1/pset01/mario-more/credit/credit.c
#include <cs50.h>
#include <stdio.h>

long extract_last_digit(long n)
{
    return n % 10;
}

long extract_left_digits(long n)
{
        return (n - extract_last_digit(n)) / 10;
}

int is_even(int n)
{
    return n % 2 == 0;
}

int main(void)
{

    //get user's input
    long number = get_long("card number: ");

    int last_two_digits = 0;    //make a variable called last two digits that we'll use below

    int i = 1, sum1 = 0, sum2 = 0;

    while (number > 0)
    {
        if (number / 10 > 0 && number / 10 < 10)
        {
            last_two_digits = number;
        }
        if (is_even(i)) //luhn's algorithm tell's us to multiply every second digit by 2 AND we get every second digit by using is_even
        {
            int p = number % 10 * 2; //if the var p is >= 10 we have to further divide it into two digits
            if (extract_left_digits(p) == 0) //if var p < 10 we add the p to sum2
            {
                sum2 += p;
            }
            else    //if p >= 10 we further divide it in two digits
            {
                sum2 += extract_left_digits(p);
                sum2 += p % 10;
            }
        }
        else    //sum1 for every second digit starting from last digit (and including the last digit)
        {
            sum1 += number % 10;
        }
        i++;
        number = extract_left_digits(number);   //putting extract_left_digits at the end of our while loop avoids skipping the last number if it were at the begining of the while loop
    }

    int number_of_digits = i - 1;   //because we're starting the counter i from 1 we need to subtract 1

    if (((sum1 + sum2) % 10 == 0) && (number_of_digits == 16 || number_of_digits == 13) && (last_two_digits / 10 == 4))
    {
        printf("VISA\n");
    }
    else if (((sum1 + sum2) % 10 == 0) && (number_of_digits == 16) && (last_two_digits == 51 || last_two_digits == 52 || last_two_digits == 53 || last_two_digits == 54 || last_two_digits == 55))
    {
        printf("MASTERCARD\n");
    }
    else if (((sum1 + sum2) % 10 == 0) && (number_of_digits == 15) && (last_two_digits == 34 || last_two_digits == 37))
    {
        printf("AMEX\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

File: week6/pset06/sentimental-credit/credit.py
import re


def main():
    number_list = []
    while True:
        num = input("Card number: ")
        if re.search(r'^[0-9]+$', num):
            break
    number = int(num)
    for i in range(len(num)):
        last = last_digit(number)
        number_list.append(last)
        number = extract_left_digits(number)
    print(number_list)
    sum = sum1(number_list) + sum2(number_list)
    print(sum)
    if len(number_list) == 16 and last_digit(sum) == 0 and number_list[15] == 5 and (number_list[14] == 1 or number_list[14] == 2 or number_list[14] == 3 or number_list[14] == 4 or number_list[14] == 5):
        print("MASTERCARD")
    elif (len(number_list) == 13 or len(number_list) == 16) and last_digit(sum) == 0 and (number_list[15] == 4 or number_list[12] == 4):
        print("VISA")
    elif len(number_list) == 15 and last_digit(sum) == 0 and number_list[14] == 3 and (number_list[13] == 4 or number_list[13] == 7):
        print("AMEX")
    else:
        print("INVALID")


def sum1(number_list):
    sum1 = 0
    for i in range(0, len(number_list), 2):
        sum1 += number_list[i]
    return sum1


def sum2(number_list):
    sum2 = 0
    for i in range(1, len(number_list), 2):
        if number_list[i] * 2 >= 10:
            bigger_num = number_list[i] * 2
            sum2 += last_digit(bigger_num) + extract_left_digits(bigger_num)
        else:
            sum2 += number_list[i] * 2
    return sum2


def last_digit(number):
    return number % 10


# Extracts left digits one at a time
def extract_left_digits(number):
    return int((number - last_digit(number)) / 10)


if __name__ == "__main__":
    main()